src/backend/gl/src/command.rs

```diff
@@ -127,47 +134,188 @@ pub const RESET: [Command; 14] = [
         offset: None,
         samples: None,
     }),
-    Command::SetViewport(Rect{x: 0, y: 0, w: 0, h: 0}),
+    Command::SetViewport(Rect {
+        x: 0,
+        y: 0,
+        w: 0,
+        h: 0,
+    }),
     Command::SetScissor(None),
     Command::SetDepthState(None),
     Command::SetStencilState(None, (0, 0), s::CullFace::Nothing),
     Command::SetBlendState(0, COLOR_DEFAULT),
     Command::SetBlendState(1, COLOR_DEFAULT),
     Command::SetBlendState(2, COLOR_DEFAULT),
     Command::SetBlendState(3, COLOR_DEFAULT),
-    Command::SetBlendColor([0f32; 4]),
+    Command::SetBlendColor([0f32; 4])
 ];
 
 struct Cache {
     primitive: gl::types::GLenum,
     index_type: c::IndexType,
+    current_vbs: Option<c::pso::VertexBufferSet<Resources>>,
     attributes: [Option<BufferElement>; c::MAX_VERTEX_ATTRIBUTES],
     resource_binds: [Option<gl::types::GLenum>; c::MAX_RESOURCE_VIEWS],
     scissor: bool,
     target_dim: (u16, u16, u16),
     stencil: Option<s::Stencil>,
-    //blend: Option<s::Blend>,
+    // blend: Option<s::Blend>,
     cull_face: s::CullFace,
     draw_mask: u32,
+
+    program: Program,
+    constant_buffer: Option<c::pso::ConstantBufferParam<Resources>>,
+    resource_view: Option<c::pso::ResourceViewParam<Resources>>,
+    scissor_test: Option<Rect>,
+    depth_state: Option<s::Depth>,
+    blend_state: Option<(c::ColorSlot, s::Color)>,
+    blend_color: Option<ColorValue>,
+    viewport: Option<Rect>,
+    rasterizer: Option<s::Rasterizer>,
+    framebuffer: Option<(Access, FrameBuffer)>,
+    index: Buffer,
 }
 
 impl Cache {
     pub fn new() -> Cache {
         Cache {
             primitive: 0,
             index_type: c::IndexType::U16,
+            current_vbs: None,
             attributes: [None; c::MAX_VERTEX_ATTRIBUTES],
             resource_binds: [None; c::MAX_RESOURCE_VIEWS],
             scissor: false,
             target_dim: (0, 0, 0),
             stencil: None,
             cull_face: s::CullFace::Nothing,
-            //blend: None,
+            // blend: None,
             draw_mask: 0,
+
+            program: 0,
+            constant_buffer: None,
+            resource_view: None,
+            scissor_test: None,
+            depth_state: None,
+            blend_state: None,
+            blend_color: None,
+            viewport: None,
+            rasterizer: None,
+            framebuffer: None,
+            index: 0,
+        }
+    }
+
+    fn push(&self, buf: &mut Vec<Command>, command: Command) {
+        buf.push(command)
+    }
+
+    fn bind_program(&mut self, buf: &mut Vec<Command>, program: Program) {
+        if program == self.program {
+            return;
+        }
+        self.program = program;
+        buf.push(Command::BindProgram(program));
+    }
+
+    fn bind_constant_buffer(&mut self, buf: &mut Vec<Command>, constant_buffer: c::pso::ConstantBufferParam<Resources>) {
+        if let Some(cb) = self.constant_buffer {
```

Comment:
nit: could be simplified to `if self.constant_buffer == Some(constant_buffer)`
if it complains about the type being moved and such, you can compare by reference `if self.constant_buffer.as_ref() == Some(&constant_buffer)`